drawBot/context/baseContext.py

```diff
@@ -886,6 +874,62 @@ def copy(self):
         return new
 
 
+def makeTextBoxes(attributedString, xy, align, plainText):
+    extraPadding = 20
+    x, y = xy
+    w, h = attributedString.size()
+    w += extraPadding
+
+    if align is not None:
+        attributedString = attributedString.mutableCopy()
+        # overwrite all align settings in each paragraph style
+        def block(value, rng, stop):
+            value = value.mutableCopy()
+            value.setAlignment_(FormattedString._textAlignMap[align])
+            attributedString.addAttribute_value_range_(AppKit.NSParagraphStyleAttributeName, value, rng)
+        attributedString.enumerateAttribute_inRange_options_usingBlock_(AppKit.NSParagraphStyleAttributeName, (0, len(attributedString)), 0, block)
+
+    setter = CoreText.CTFramesetterCreateWithAttributedString(attributedString)
+    path = Quartz.CGPathCreateMutable()
+    Quartz.CGPathAddRect(path, None, Quartz.CGRectMake(x, y, w, h * 2))
+    frame = CoreText.CTFramesetterCreateFrame(setter, (0, 0), path, None)
+    ctLines = CoreText.CTFrameGetLines(frame)
+    origins = CoreText.CTFrameGetLineOrigins(frame, (0, len(ctLines)), None)
+    boxes = []
+
+    for i, (originX, originY) in enumerate(origins):
+        ctLine = ctLines[i]
+        rng = CoreText.CTLineGetStringRange(ctLine)
+
+        attributedSubstring = attributedString.attributedSubstringFromRange_(rng)
+        # strip trailing returns
+        if attributedSubstring.string()[-1] in ["\n", "\r"]:
+            rng.length -= 1
+            attributedSubstring = attributedString.attributedSubstringFromRange_(rng)
```

Comment:
How about something like:
`attributedString.deleteCharactersInRange_((rng.length - 1, 1))`
It's a mutable attr string, and I would _assume_ that `attributedSubstringFromRange_()` earlier gave you a copy. (I will verify that.)

Reply:
Yup, all good:
```python
s = "abcdefghijklmnop"
fs = FormattedString()
fs.append(s)
ats = fs.getNSObject()
assert ats.string() == s
sub = ats.attributedSubstringFromRange_((3, 3))
assert sub.string() == "def"
sub.deleteCharactersInRange_((1, 1))
assert sub.string() == "df"
assert ats.string() == s
```

Reply:
changed